src/HotChocolate/Data/src/Data/Projections/Expressions/LevelOperations.cs

```diff
@@ -0,0 +1,46 @@
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Linq.Expressions;
+
+namespace HotChocolate.Data.Projections.Expressions;
+
+public sealed class LevelOperations : IEnumerable<Expression>
+{
+    private readonly List<Expression> _operations = new();
+
+    // This is to support the queue behavior.
+    private int _startIndex = 0;
+
+    public void Enqueue(Expression expression) => _operations.Add(expression);
+    public Expression Dequeue() => _operations[_startIndex++];
+    public Expression Peek() => _operations[_startIndex];
+    public bool TryPeek(out Expression expression)
+    {
+        if (_startIndex < _operations.Count)
+        {
+            expression = _operations[_startIndex];
+            return true;
+        }
+
+        expression = default!;
+        return false;
+    }
+    public int Count => _operations.Count - _startIndex;
+
+    public Expression this[int index]
```

Comment:
I guess i know why you want to add this. Do you want to rewrite stuff from previous visitations? 
If so this sometimes seem like a good idea, but you will end up with a ton of issues. Its important that you only modify the "current" thing you are visiting via enter and leave. 
if you start modifying other things then you will very likely run into issues when you visit more complex queries. 
The stack/queue semantics is actually a pretty good way to help enforcing it.

If i made the wrong assumptions here let me know. :)

Reply:
Yeah well, the thing is, the fact that the level did not allow overwriting existing expressions made me override `ProjectionFieldHandler` instead of just defining an interceptor, and then also copy-pasting most of the code from the implementation. I guess all I actually needed was not even an index accessor, but just an ability to update the last added element. But I don't see why updating the expressions by index wouldn't be allowed either.

> if you start modifying other things then you will very likely run into issues when you visit more complex queries.

If you're interacting with the projection engine at this level, you better know what you're doing. In fact, you have to know. So this is not an issue.

Reply:
> If you're interacting with the projection engine at this level, you better know what you're doing. In fact, you have to know. So this is not an issue.

Its because you will make mistakes this way. In a visitor there is usually only one place to add something. If you are able to modify the state to much, there is more than one place. The visitation is a easy concept. Enter/Leave. As long as you work in these boundaries and it works once, it also works for complex trees. If you leave these boundaries, you can easily have side effects.

Reply:
Sure, then I guess it's just that more useful abstractions are needed.